bin/src/prometheus_server.rs

```diff
@@ -0,0 +1,83 @@
+use std::future::{Ready, ready};
+
+use hyper::{Request, Response, body::Incoming, service::Service};
+use hyper_util::{
+    rt::{TokioExecutor, TokioIo},
+    server::{conn::auto::Builder, graceful::GracefulShutdown},
+};
+use metrics_exporter_prometheus::PrometheusHandle;
+use tokio::{net::TcpListener, select, task::JoinHandle};
+use tokio_util::sync::CancellationToken;
+use tracing::debug;
+
+/// Runs an HTTP server that responds to Prometheus scrape requests.
+pub fn prometheus_server(
+    handle: PrometheusHandle,
+    listener: TcpListener,
+) -> (JoinHandle<()>, CancellationToken) {
+    let service = PrometheusService::new(handle);
+    let token = CancellationToken::new();
+    let token_clone = token.clone();
+    let shutdown = GracefulShutdown::new();
+    let handle = tokio::spawn(async move {
+        let builder = Builder::new(TokioExecutor::new());
+        loop {
+            let stream = select! {
+                result = listener.accept() => {
+                    match result {
+                        Ok((stream, _)) => stream,
+                        Err(error) => {
+                            debug!(%error, "error accepting connection");
+                            continue;
+                        }
+                    }
+                },
+                _ = token.cancelled() => {
+                    drop(listener);
+                    break;
+                },
+            };
+            let io = TokioIo::new(stream);
+            let conn = builder.serve_connection_with_upgrades(io, service.clone());
+            let conn = shutdown.watch(conn.into_owned());
+            tokio::spawn(async move {
+                if let Err(error) = conn.await {
+                    debug!(%error, "connection error");
+                }
+            });
+        }
+        shutdown.shutdown().await;
+    });
+    (handle, token_clone)
+}
+
+#[derive(Clone)]
+struct PrometheusService {
+    handle: PrometheusHandle,
+}
+
+impl PrometheusService {
+    fn new(handle: PrometheusHandle) -> Self {
+        Self { handle }
+    }
+}
+
+impl Service<Request<Incoming>> for PrometheusService {
+    type Response = Response<String>;
+
+    type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+    type Future =
+        Ready<Result<Response<String>, Box<dyn std::error::Error + Send + Sync + 'static>>>;
+
+    fn call(&self, _req: Request<Incoming>) -> Self::Future {
+        let body = self.handle.render();
+        let response_builder =
+            Response::builder().header("Content-Type", "text/plain; version=0.0.4");
+        let response = match response_builder.body(body) {
+            Ok(response) => response,
+            Err(e) => return ready(Err(Box::new(e))),
+        };
+        ready(Ok(response))
```

Comment:
Is this copy/pasted from upstream? Suggest:

- Using the `CONTENT_TYPE` const from the http crate
- Inlining the `body` and `response` bindings

Reply:
Not directly, `metrics-exporter-prometheus` uses `service_fn()`, for example.